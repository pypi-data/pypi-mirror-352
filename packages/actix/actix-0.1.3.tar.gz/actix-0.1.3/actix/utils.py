# actix/utils.py
# Shared utility functions, if any, can be placed here.
# For now, it can remain empty.
