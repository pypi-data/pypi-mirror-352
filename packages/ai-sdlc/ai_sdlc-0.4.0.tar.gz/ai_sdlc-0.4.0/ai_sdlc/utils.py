"""Shared helpers."""

from __future__ import annotations

from pathlib import Path
import json
import re
import sys
import unicodedata

def find_project_root() -> Path:
    """Find project root by searching for .aisdlc file in current and parent directories."""
    current_dir = Path.cwd()
    for parent in [current_dir] + list(current_dir.parents):
        if (parent / ".aisdlc").exists():
            return parent
    # Fallback or error
    print("Error: .aisdlc not found. Ensure you are in an ai-sdlc project directory.")
    sys.exit(1)

ROOT = find_project_root()

# --- TOML loader (Python ≥3.11 stdlib) --------------------------------------
try:
    import tomllib as _toml  # Python 3.11+
except ModuleNotFoundError:  # pragma: no cover – fallback for < 3.11
    import tomli as _toml    # noqa: D401  # `uv pip install tomli`


def load_config() -> dict:
    cfg_path = ROOT / ".aisdlc"
    if not cfg_path.exists():
        raise FileNotFoundError(".aisdlc manifest missing – run `aisdlc init`.")
    try:
        return _toml.loads(cfg_path.read_text())
    except _toml.TOMLDecodeError as e:
        print(f"❌ Error: '.aisdlc' configuration file is corrupted: {e}")
        print("Please fix the .aisdlc file or run 'aisdlc init' in a new directory.")
        import sys
        sys.exit(1)


def slugify(text: str) -> str:
    """kebab-case ascii only"""
    slug = unicodedata.normalize("NFKD", text).encode("ascii", "ignore").decode()
    slug = re.sub(r"[^a-zA-Z0-9]+", "-", slug).strip("-").lower()
    return slug or "idea"


def read_lock() -> dict:
    path = ROOT / ".aisdlc.lock"
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text())
    except json.JSONDecodeError:
        print("⚠️  Warning: '.aisdlc.lock' file is corrupted or not valid JSON. Treating as empty.")
        return {}


def write_lock(data: dict) -> None:
    (ROOT / ".aisdlc.lock").write_text(json.dumps(data, indent=2))
