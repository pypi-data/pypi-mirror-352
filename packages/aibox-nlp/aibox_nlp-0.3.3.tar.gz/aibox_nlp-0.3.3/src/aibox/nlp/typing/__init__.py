"""Definições comuns de type aliases."""

from typing import Callable, TypeAlias

import numpy as np
import torch

#: Define a noção de array-like numéricos.
ArrayLike: TypeAlias = (
    list[int]
    | list[float]
    | np.ndarray[np.int32]
    | np.ndarray[np.float32]
    | torch.Tensor
)

#: Define array-like de textos.
TextArrayLike: TypeAlias = list[str] | np.ndarray[np.str_]

#: Define uma função de pós-processamento.
PostProcessing: TypeAlias = Callable[[np.ndarray], np.ndarray]
