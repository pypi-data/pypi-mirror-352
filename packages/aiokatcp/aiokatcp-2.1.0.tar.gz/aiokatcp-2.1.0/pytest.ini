[pytest]
markers =
    server_cls: internal use
    channel_cls: internal use
    channel_args: internal use
asyncio_mode = auto
