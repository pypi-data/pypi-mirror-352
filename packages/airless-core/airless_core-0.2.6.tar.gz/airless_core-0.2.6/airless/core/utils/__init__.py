from .config import (get_config)
from .enum import (BaseEnum)

__all__ = [
    'get_config',
    'BaseEnum'
]
