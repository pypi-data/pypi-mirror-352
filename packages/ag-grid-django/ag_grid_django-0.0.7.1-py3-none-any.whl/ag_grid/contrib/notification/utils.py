from django.utils.encoding import force_str

from asgiref.sync import async_to_sync
from channels.layers import get_channel_layer


def send_notification_to_user(user_id, notification_data):
    """
    Send notification to a specific user

    Args:
        user_id: ID of the target user
        notification_data: Dictionary containing notification details
    """
    channel_layer = get_channel_layer()
    if not channel_layer:
        return
    async_to_sync(channel_layer.group_send)(f"user_{user_id}_notifications", {"type": "notification", "notification": notification_data})


def send_notification_to_all_staff(notification_data):
    """
    Send notification to all staff users

    Args:
        notification_data: Dictionary containing notification details
    """
    channel_layer = get_channel_layer()
    if not channel_layer:
        return
    async_to_sync(channel_layer.group_send)("all_staff_notifications", {"type": "notification", "notification": notification_data})


def send_notification(notification):
    """
    Send notification object through WebSocket

    Args:
        notification: Notification model instance
    """
    # Convert notification to serializable format
    notification_data = {
        "id": notification.id,
        "type": force_str(notification.type),
        "title": force_str(notification.title),
        "message": force_str(notification.message),
        "action_url": force_str(notification.action_url) if notification.action_url else None,
        "file_url": force_str(notification.file_url) if notification.file_url else None,
        "created_at": notification.created_at.isoformat(),
    }

    # Send to appropriate targets
    if notification.target_type == "specific_user" and notification.user:
        send_notification_to_user(notification.user.id, notification_data)
    else:
        send_notification_to_all_staff(notification_data)
