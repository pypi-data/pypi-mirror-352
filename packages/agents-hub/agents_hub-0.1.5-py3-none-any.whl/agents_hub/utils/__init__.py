"""
Utility functions and classes for the Agents Hub framework.

This module provides various utilities used across the framework.
"""

from agents_hub.utils.approval import ApprovalInterface
from agents_hub.utils.json_parser import (
    RobustJSONParser,
    extract_json,
    JSONParsingError,
)

__all__ = [
    "ApprovalInterface",
    "RobustJSONParser",
    "extract_json",
    "JSONParsingError",
]
