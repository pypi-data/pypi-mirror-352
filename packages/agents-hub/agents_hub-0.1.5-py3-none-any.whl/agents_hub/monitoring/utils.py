"""
Utility functions for the monitoring system.
"""

from typing import Dict, List, Any, Optional, Union, Tuple
import json
import logging

# Initialize logger
logger = logging.getLogger(__name__)

# Try to import tiktoken for OpenAI token counting
try:
    import tiktoken

    TIKTOKEN_AVAILABLE = True
except ImportError:
    TIKTOKEN_AVAILABLE = False
    logger.warning(
        "tiktoken is not installed. Token counting for OpenAI models will be approximate. "
        "Install tiktoken for accurate token counting: pip install tiktoken"
    )


def count_tokens(text: str, model: str) -> int:
    """
    Count the number of tokens in a text string for a specific model.

    Args:
        text: Text to count tokens for
        model: Model name to use for token counting

    Returns:
        Number of tokens
    """
    if not text:
        return 0

    # Use tiktoken for OpenAI models if available
    if TIKTOKEN_AVAILABLE and model.startswith(("gpt-", "text-davinci-")):
        try:
            encoding = tiktoken.encoding_for_model(model)
            return len(encoding.encode(text))
        except Exception as e:
            logger.warning(f"Error using tiktoken for model {model}: {e}")

    # Fallback to approximate token counting
    # A rough approximation is 4 characters per token for English text
    return len(text) // 4


def count_tokens_for_messages(
    messages: List[Dict[str, Any]], model: str
) -> Dict[str, int]:
    """
    Count the number of tokens in a list of messages for a specific model.

    Args:
        messages: List of messages to count tokens for
        model: Model name to use for token counting

    Returns:
        Dictionary with token counts (input_tokens)
    """
    if not messages:
        return {"input_tokens": 0}

    # For OpenAI chat models with tiktoken
    if TIKTOKEN_AVAILABLE and model.startswith(("gpt-")):
        try:
            encoding = tiktoken.encoding_for_model(model)

            # From OpenAI's documentation:
            # Every message follows <|start|>{role/name}\n{content}<|end|>
            # If there's a name, the role is omitted
            num_tokens = 0
            for message in messages:
                num_tokens += 4  # <|start|> and <|end|> tokens
                for key, value in message.items():
                    num_tokens += len(encoding.encode(str(value)))
                    if key == "name":  # If there's a name, the role is omitted
                        num_tokens -= 1  # Role is omitted

            num_tokens += 2  # Every reply is primed with <|start|>assistant<|message|>

            return {"input_tokens": num_tokens}
        except Exception as e:
            logger.warning(f"Error using tiktoken for model {model}: {e}")

    # Fallback to approximate token counting
    total_tokens = 0
    for message in messages:
        content = message.get("content", "")
        if isinstance(content, str):
            total_tokens += count_tokens(content, model)

    return {"input_tokens": total_tokens}


def get_model_pricing(model: str, provider: str) -> Tuple[float, float]:
    """
    Get the pricing information for a specific model and provider.

    Args:
        model: Model name
        provider: Provider name

    Returns:
        Tuple of (input_price_per_1k_tokens, output_price_per_1k_tokens)
    """
    # OpenAI pricing (as of April 2025)
    openai_pricing = {
        "gpt-4o": (5.0, 15.0),  # $5.00 per 1M input tokens, $15.00 per 1M output tokens
        "gpt-4o-mini": (
            0.15,
            0.60,
        ),  # $0.15 per 1M input tokens, $0.60 per 1M output tokens
        "gpt-4": (
            30.0,
            60.0,
        ),  # $30.00 per 1M input tokens, $60.00 per 1M output tokens
        "gpt-4-turbo": (
            10.0,
            30.0,
        ),  # $10.00 per 1M input tokens, $30.00 per 1M output tokens
        "gpt-3.5-turbo": (
            0.5,
            1.5,
        ),  # $0.50 per 1M input tokens, $1.50 per 1M output tokens
    }

    # Claude pricing (as of April 2025)
    claude_pricing = {
        "claude-3-opus-20240229": (
            15.0,
            75.0,
        ),  # $15.00 per 1M input tokens, $75.00 per 1M output tokens
        "claude-3-sonnet-20240229": (
            3.0,
            15.0,
        ),  # $3.00 per 1M input tokens, $15.00 per 1M output tokens
        "claude-3-haiku-20240307": (
            0.25,
            1.25,
        ),  # $0.25 per 1M input tokens, $1.25 per 1M output tokens
    }

    # Gemini pricing (as of April 2025)
    gemini_pricing = {
        "gemini-1.5-pro": (
            7.0,
            21.0,
        ),  # $7.00 per 1M input tokens, $21.00 per 1M output tokens
        "gemini-1.5-flash": (
            0.35,
            1.05,
        ),  # $0.35 per 1M input tokens, $1.05 per 1M output tokens
    }

    # Default pricing if model not found
    default_pricing = (
        10.0,
        30.0,
    )  # $10.00 per 1M input tokens, $30.00 per 1M output tokens

    # Get pricing based on provider and model
    if provider.lower() == "openai" or provider.lower() == "openaiprovider":
        # Try exact match first
        if model in openai_pricing:
            return openai_pricing[model]

        # Try prefix match
        for model_prefix, pricing in openai_pricing.items():
            if model.startswith(model_prefix):
                return pricing

        # Handle specific model variants
        if "gpt-4o" in model.lower():
            return openai_pricing["gpt-4o"]
        elif "gpt-4" in model.lower():
            return openai_pricing["gpt-4"]
        elif "gpt-3.5" in model.lower():
            return openai_pricing["gpt-3.5-turbo"]

    elif provider.lower() in ["anthropic", "claude", "claudeprovider"]:
        # Try exact match first
        if model in claude_pricing:
            return claude_pricing[model]

        # Try prefix match
        for model_prefix, pricing in claude_pricing.items():
            if model.startswith(model_prefix):
                return pricing

        # Handle specific model variants
        if "opus" in model.lower():
            return claude_pricing["claude-3-opus-20240229"]
        elif "sonnet" in model.lower():
            return claude_pricing["claude-3-sonnet-20240229"]
        elif "haiku" in model.lower():
            return claude_pricing["claude-3-haiku-20240307"]

    elif provider.lower() in ["google", "gemini", "geminiprovider"]:
        # Try exact match first
        if model in gemini_pricing:
            return gemini_pricing[model]

        # Try prefix match
        for model_prefix, pricing in gemini_pricing.items():
            if model.startswith(model_prefix):
                return pricing

        # Handle specific model variants
        if "pro" in model.lower():
            return gemini_pricing["gemini-1.5-pro"]
        elif "flash" in model.lower():
            return gemini_pricing["gemini-1.5-flash"]

    # Return default pricing if no match found
    logger.warning(
        f"No pricing information found for {provider}/{model}. Using default pricing."
    )
    return default_pricing


def estimate_cost(
    provider: str, model: str, input_tokens: int, output_tokens: int
) -> float:
    """
    Estimate the cost of an LLM call based on token usage.

    Args:
        provider: Provider name
        model: Model name
        input_tokens: Number of input tokens
        output_tokens: Number of output tokens

    Returns:
        Estimated cost in USD
    """
    # Get pricing information
    input_price_per_1m, output_price_per_1m = get_model_pricing(model, provider)

    # Convert to price per token
    input_price_per_token = input_price_per_1m / 1_000_000
    output_price_per_token = output_price_per_1m / 1_000_000

    # Calculate cost
    input_cost = input_tokens * input_price_per_token
    output_cost = output_tokens * output_price_per_token
    total_cost = input_cost + output_cost

    return total_cost
