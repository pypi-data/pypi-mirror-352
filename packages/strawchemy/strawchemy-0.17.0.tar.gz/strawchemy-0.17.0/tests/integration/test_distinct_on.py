from __future__ import annotations

from typing import TYPE_CHECKING

import pytest

from syrupy.assertion import SnapshotAssertion
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .fixtures import QueryTracker
from .typing import RawRecordData

if TYPE_CHECKING:
    from strawchemy import StrawchemyConfig


@pytest.fixture
def raw_colors() -> RawRecordData:
    return [
        {"id": 1, "name": "Red"},
        {"id": 2, "name": "Red"},
        {"id": 3, "name": "Orange"},
        {"id": 4, "name": "Orange"},
        {"id": 5, "name": "Pink"},
    ]


@pytest.mark.parametrize(
    "deterministic_ordering",
    [pytest.param(True, id="deterministic-ordering"), pytest.param(False, id="non-deterministic-ordering")],
)
@pytest.mark.snapshot
async def test_distinct_on(
    any_query: AnyQueryExecutor,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    config: StrawchemyConfig,
    deterministic_ordering: bool,
) -> None:
    config.deterministic_ordering = deterministic_ordering
    result = await maybe_async(any_query("{ colors(distinctOn: [name]) { id name } }"))
    assert not result.errors
    assert result.data

    expected = [{"id": 1, "name": "Red"}, {"id": 3, "name": "Orange"}, {"id": 5, "name": "Pink"}]
    assert all(color in result.data["colors"] for color in expected)

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_distinct_and_order_by(
    any_query: AnyQueryExecutor, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    result = await maybe_async(
        any_query("{ colors(distinctOn: [name], orderBy: [{name: ASC}, {id: DESC}]) { id name } }")
    )
    assert not result.errors
    assert result.data

    expected = [{"id": 2, "name": "Red"}, {"id": 4, "name": "Orange"}, {"id": 5, "name": "Pink"}]
    assert all(color in result.data["colors"] for color in expected)

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
