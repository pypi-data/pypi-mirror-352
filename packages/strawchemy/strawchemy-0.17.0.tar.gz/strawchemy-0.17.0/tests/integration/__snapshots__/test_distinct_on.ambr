# serializer version: 1
# name: test_distinct_and_order_by[session-tracked-async-aiosqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_and_order_by[session-tracked-async-asyncmy_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %s
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_and_order_by[session-tracked-async-asyncpg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = $1::INTEGER
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_and_order_by[session-tracked-async-psycopg_async_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_and_order_by[session-tracked-sync-psycopg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_and_order_by[session-tracked-sync-sqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 color.name AS name__1,
                 color.id AS id__1,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.name ASC, color.id DESC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.name ASC,
                    color.id DESC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
   ORDER BY color.name ASC,
            color.id DESC
  '''
# ---
# name: test_distinct_on[session-tracked-async-asyncmy_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %s
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-asyncpg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = $1::INTEGER
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-deterministic-ordering-aiosqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-deterministic-ordering-asyncmy_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %s
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-deterministic-ordering-asyncpg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = $1::INTEGER
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-deterministic-ordering-psycopg_async_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-async-non-deterministic-ordering-aiosqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
  '''
# ---
# name: test_distinct_on[session-tracked-async-non-deterministic-ordering-asyncmy_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %s
  '''
# ---
# name: test_distinct_on[session-tracked-async-non-deterministic-ordering-asyncpg_engine]
  '''
  SELECT DISTINCT
      ON (color.name) color.name,
         color.id
    FROM color AS color
  '''
# ---
# name: test_distinct_on[session-tracked-async-non-deterministic-ordering-psycopg_async_engine]
  '''
  SELECT DISTINCT
      ON (color.name) color.name,
         color.id
    FROM color AS color
  '''
# ---
# name: test_distinct_on[session-tracked-async-psycopg_async_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-sync-deterministic-ordering-psycopg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-sync-deterministic-ordering-sqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
   ORDER BY color.id ASC
  '''
# ---
# name: test_distinct_on[session-tracked-sync-non-deterministic-ordering-psycopg_engine]
  '''
  SELECT DISTINCT
      ON (color.name) color.name,
         color.id
    FROM color AS color
  '''
# ---
# name: test_distinct_on[session-tracked-sync-non-deterministic-ordering-sqlite_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = ?
  '''
# ---
# name: test_distinct_on[session-tracked-sync-psycopg_engine]
  '''
  SELECT color.name,
         color.id
    FROM (
          SELECT color.id AS id,
                 color.name AS name,
                 row_number() OVER (PARTITION BY color.name ORDER BY color.id ASC) AS __strawchemy_distinct_on_rank_0
            FROM color AS color
           ORDER BY color.id ASC
         ) AS color
   WHERE color.__strawchemy_distinct_on_rank_0 = %(param_1)s::INTEGER
   ORDER BY color.id ASC
  '''
# ---
